doc/source/notebooks/advanced/variational_fourier_features.pct.py

```diff
@@ -230,97 +230,105 @@ def gauss_kl_vff(q_mu, q_sqrt, K):
 
 
 # %%
-@gpflow.conditionals.conditional.register(
-    TensorLike, FourierFeatures1D, gpflow.kernels.Kernel, TensorLike
-)
-def conditional_vff(
-    Xnew,
-    inducing_variable,
-    kernel,
-    f,
-    *,
-    full_cov=False,
-    full_output_cov=False,
-    q_sqrt=None,
-    white=False,
-):
-    """
-     - Xnew are the points of the data or minibatch, size N x D (tf.array, 2d)
-     - feat is an instance of features.InducingFeature that provides `Kuu` and `Kuf` methods
-       for Fourier features, this contains the limits of the bounding box and the frequencies
-     - f is the value (or mean value) of the features (i.e. the weights)
-     - q_sqrt (default None) is the Cholesky factor of the uncertainty about f
-       (to be propagated through the conditional as per the GPflow inducing-point implementation)
-     - white (defaults False) specifies whether the whitening has been applied
-
-    Given the GP represented by the inducing points specified in `feat`, produce the mean and
-    (co-)variance of the GP at the points Xnew.
-
-       Xnew :: N x D
-       Kuu :: M x M
-       Kuf :: M x N
-       f :: M x K, K = 1
-       q_sqrt :: K x M x M, with K = 1
-    """
-    if full_output_cov:
-        raise NotImplementedError
-
-    # num_data = tf.shape(Xnew)[0]  # M
-    num_func = tf.shape(f)[1]  # K
-
-    Kuu = cov.Kuu(inducing_variable, kernel)  # this is now a LinearOperator
-    Kuf = cov.Kuf(inducing_variable, kernel, Xnew)  # still a Tensor
-
-    KuuInv_Kuf = Kuu.solve(Kuf)
-
-    # compute the covariance due to the conditioning
-    if full_cov:
-        fvar = kernel(Xnew) - tf.matmul(Kuf, KuuInv_Kuf, transpose_a=True)
-        shape = (num_func, 1, 1)
-    else:
-        KufT_KuuInv_Kuf_diag = tf.reduce_sum(Kuf * KuuInv_Kuf, axis=-2)
-        fvar = kernel(Xnew, full_cov=False) - KufT_KuuInv_Kuf_diag
-        shape = (num_func, 1)
-    fvar = tf.expand_dims(fvar, 0) * tf.ones(
-        shape, dtype=gpflow.default_float()
-    )  # K x N x N or K x N
+import gpflow.posterior
+
+
+class VFFPosterior(gpflow.posterior.AbstractPosterior):
+    def _precompute(self):
+        """
+        Not implemented yet
+        """
+        return np.nan, np.nan
+
+    def _conditional_with_precompute(self, Xnew, full_cov, full_output_cov):
```

Comment:
Can this method be removed by renaming `_conditional_fused` to `_conditional_with_precompute`?

Reply:
Same question here. I don't think the name is important. I would like to know the reasoning for having two methods that do the same job.

Reply:
They do the same job from a mathematical point of view, but they don't do the same job from a finite-precision computational numerics point of view. When training, the fused conditional requires less ops than with the precompute. When testing, the one with precompute has significant advantages. Having both addresses @markvdw's concerns about completely changing the numerics at the core of GPflow.